### ISSUE TITLE
getAttribute("value") returns empty String for UIAButton in iOS; works for other elements and other attributes

### ISSUE THREAD

**Issue by ChiWhiteSox56** [appium/appium#6710]
## The problem

The getAttribute("value") method returns an empty String for UIAButton elements in iOS. This only happens when running my automated tests; if I launch Appium via the UI and launch the Inspector, the "value" property can be seen and is populated.

## Environment

- Appium server version: Version 1.5.3 (Ara)
- Desktop OS/version used to run Appium if necessary: OS X 10.11.5
- Node.js version: v6.3.0
- Mobile platform/version under test: iOS 9.3
- Real device or emulator/simulator: Simulator

## Details

I am working on automated tests for an app that has buttons that use a change of imagery to show when the button is selected/deselected. As such, I can not use the isSelected() method to check whether the buttons are selected. I worked with the app developer to get "selected" and "deselected" values added to the element as appropriate. When I locate the element and use .getAttribute("value"), I get an empty String. However, I use getAttribute("value") elsewhere in the code and it works as expected. Examples:

**This works:** 
return driver.findElement(NOTES_TEXT_FIELD).getAttribute("value").toString();
- NOTES_TEXT_FIELD element type: UIATextView
- element is located by xpath
- didn't work until I added ".toString()

**This does not work:** 
return driver.findElement(BTN_A).getAttribute("value").toString();
- BTN_A element type: UIAButton
- element is located by AccessibilityId; changed locator from MobileBy.AccessibilityId() to By.xpath() but it didn't fix the problem
- tried with and without .toString(); doesn't work either way
- if I look at the element with either the Appium inspector or the Accessibility Inspector, the "selected" and "deselected" traits added by the developer are present under "value"

I tried adding `driver.getPageSource()` immediately before trying to call getAttribute("value") in hopes that it might refresh the XML and make the value visible, but the output shows that it did not help: 
`<UIAButton name="A" label="A" value="" dom="" enabled="true" valid="true" visible="true" hint="" path="/0/0/20/0/0/2/1" x="20" y="233.361328125" width="40" height="40">`

I should also mention that `return driver.findElement(BTN_A).getAttribute("label")` does return 'A' as expected, which makes this issue even more puzzling. It seems like there is a problem with UIAButton and the "value" attribute specifically. Based on what I've read on other forums, this may even be isolated to UIAButtons that are contained within a UIATableView.

I posted this issue in Appium Discuss. Please see this link for screenshots of the Appium Inspector and Accessibiity Inspector: https://discuss.appium.io/t/ios-getattribute-value-works-for-some-elements-returns-empty-string-for-others/11318/20 (Here, you can see that the value does show up in the Inspectors)

## Link to Appium logs

https://gist.github.com/ChiWhiteSox56/dea6c0bfd32a34a014f1d983d0895fb0

## Code To Reproduce Issue [ Good To Have ]

```
public void verifyOnlyOneAnswerCanBeSelected() throws InterruptedException {
	quizSettingsPage.createQuizWithDefaultSettings();
	// wait for answer choice A to be clickable
	quizInterfacePage.selectAnswerChoice(AnswerChoice.A);
	driver.getPageSource();
	//System.out.println(driver.getPageSource());
	quizInterfacePage.answerChoiceIsSelected(AnswerChoice.A);
}

public void answerChoiceIsSelected(AnswerChoice letter) {

	switch (letter) {
		case A:
        // debugging
		System.out.println("ATTRIBUTE: " + driver.findElement(BTN_A).getAttribute("value"));
        // debugging
		System.out.println("IS DISPLAYED: " + driver.findElement(BTN_A));
		driver.findElement(BTN_A).getAttribute("value");
		break;
			
		case B:
		driver.findElement(BTN_B).click();
		break;
			
		case C:
		driver.findElement(BTN_C).click();
		break;
			
		case D:
		driver.findElement(BTN_D).click();
		break;
	}
}
```

**Comment by mykola-mokhnach** [comment]
Probably, the testing code works too fast and gets the actual value before it is changed in code? What if you just replace 
```java
driver.findElement(BTN_A).getAttribute("value"))
```
with
```java
WebElement el = driver.findElement(BTN_A);
long msStarted = System.currentTimeMillis()
while (!el.getAttribute("value").equals("blabla") && System.currentTimeMillis() - msStarted < 5 * 1000) {
    Thread.sleep(500);
    // just to make sure the UI tree has been refreshed
    // driver.getPageSource();
}
assert (!el.getAttribute("value").equals("blabla")) : "Current button state is not equal to the expected one after 5 seconds timeout";
}
```

Also, if this still does not work (which looks more like instruments bug in that case), then you can try to change accessibility id in the app code instead of value field.

**Comment by ChiWhiteSox56** [comment]
Thank you so much for your response, @mykola-mokhnach 

I employed your solution and I still wasn't getting a value, so I decided to comb through every part of the code and double-check everything, regardless of my level of certainty about it. Sure enough, I was pointing to an older version of the app that did not have the selected/deselected value added to it. Not sure how I haven't noticed that until now, but I sincerely apologize for submitting an issue that's a result of my own error. I'll close this issue. And again, thank you so much for your time.
